OCI KMS auto-unseal plugin

vuhpham-oracle opened the pull request:
Hello Vault, I am, on behalf of Oracle Cloud Infrastructure, submitting this auto-unseal plugin using our OCI KMS service. 
Details on OCI KMS Service: https://docs.cloud.oracle.com/iaas/Content/KeyManagement/Concepts/keyoverview.htm

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault?pullRequest=6950)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault?pullRequest=6950)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault?pullRequest=6950).</sub>

jefferai commented:
Thanks for submitting this! Can you please remove all of the vendored files from this PR? It makes it hard to review as it takes forever to load up. We can fix up vendoring later.

vuhpham-oracle commented:
@jefferai I removed the vendor package. Thanks.

kalafut commented on go.mod:
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/Azure/azure-sdk-for-go v29.0.0+incompatible
 	github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78 // indirect
 	github.com/Azure/go-autorest v11.7.1+incompatible
+	github.com/Bowery/prompt v0.0.0-20190419144237-972d0ceb96f5 // indirect
```
Run `go mod tidy`

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+
+import (
+	"context"
+	b64 "encoding/base64"
```
I'd remove this. `base64.*` is pervasive in Vault, and seeing b64 in code could cause confusing, miss grep matches, etc.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+	// OCI KMS key ID to use for encryption and decryption
+	EnvOCIKMSCryptoEndpoint = "VAULT_OCIKMS_CRYPTO_ENDPOINT"
+	// Should retry for calls to KMS endpoints
+	KMSShouldRetryOperation = true
```
Suggest removing this, and the one conditional where it is referenced.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+		metrics.SetGauge(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigKeyID)
+	}
+	k.logger.Info(fmt.Sprintf("got %s: %s", KMSConfigKeyID, k.keyID))
```
I don't think we want the KMS details showing up in logs. If we do keep this, it would be cleaner like:
`k.logger.Info("OCI kms configuration", KMSConfigKeyID, k.keyID)`

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+		metrics.SetGauge(metricEncryptFailed, 1)
+		return nil, errwrap.Wrapf("error encrypting data: {{err}}", err)
+	}
+	k.logger.Info(fmt.Sprintf("successfully encrypted Vault unseal key. Storing encrypted key to storage"))
```
The encrypted data needn't be the Vault unseal key, nor do we know if anything is being stored.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+		metrics.SetGauge(metricDecryptFailed, 1)
+		return nil, errwrap.Wrapf("error base64 decrypting data: {{err}}", err)
+	}
+	k.logger.Info(fmt.Sprintf("successfully decrypted Vault unseal key"))
```
The decrypted data needn't be the Vault unseal key.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+
+// Request metadata includes retry policy
+func (k *OCIKMSSeal) getRequestMetadata() common.RequestMetadata {
+	if !KMSShouldRetryOperation {
```
As mentioned above, I think this can go away.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,305 @@
+
+	// Only retry for 5xx errors
+	retryOn5xxFunc := func(r common.OCIOperationResponse) bool {
+		return !(r.Error == nil && r.Response.HTTPResponse().StatusCode < 500)
```
```suggestion
		return r.Error != nil && r.Response.HTTPResponse().StatusCode >= 500
```
I think the non-negated version is more intuitive to read.

vuhpham-oracle replied:
Should it be || instead of && in your suggestion?

kalafut commented:
@vuhpham-oracle  Hi Vu. Looks good. Other than some minor comments, my main question is the same one I brought up on the objectstore PR regarding SetGauge for errors. We'll discuss there and the outcome will probably apply here.

vuhpham-oracle commented:
Hi @kalafut, thanks for your feedbacks! I addressed your comments in latest commit. For Gauge metric for error, we'll discuss on OCI Object Storage PR.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -182,16 +180,9 @@ func (k *OCIKMSSeal) getOCIKMSClient() (*keymanagement.KmsCryptoClient, error) {
 	// Only retry for 5xx errors
 	retryOn5xxFunc := func(r common.OCIOperationResponse) bool {
-		return !(r.Error == nil && r.Response.HTTPResponse().StatusCode < 500)
+		return r.Error != nil || r.Response.HTTPResponse().StatusCode >= 500
```
Should be `&&`, right?  We have an error, and it's code is a 5xx.

vuhpham-oracle replied:
Yes, I thought only 5xx has r.Error != nill, double checked and && is correct

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -259,7 +250,7 @@ func (k *OCIKMSSeal) Encrypt(ctx context.Context, plaintext []byte) (*physical.E
 		return nil, errwrap.Wrapf("error encrypting data: {{err}}", err)
 	}
-	k.logger.Info(fmt.Sprintf("successfully encrypted Vault unseal key. Storing encrypted key to storage"))
+	k.logger.Debug(fmt.Sprintf("successfully encrypted"))
```
`fmt.Sprintf` is unnecessary

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -294,12 +285,12 @@ func (k *OCIKMSSeal) Decrypt(ctx context.Context, in *physical.EncryptedBlobInfo
 		return nil, errwrap.Wrapf("error base64 decrypting data: {{err}}", err)
 	}
-	k.logger.Info(fmt.Sprintf("successfully decrypted Vault unseal key"))
+	k.logger.Debug(fmt.Sprintf("successfully decrypted"))
```
`fmt.Sprintf` is unnecessary

kalafut commented:
Just a couple of tweaks and this one is good (except for the Gauge discussion).

vuhpham-oracle commented:
@kalafut Thanks a lot for your feedback! We finished resolving your first round feedback. Please let us know if you have any further feedback.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -225,40 +233,48 @@ func (k *OCIKMSSeal) Encrypt(ctx context.Context, plaintext []byte) (*physical.E
 		return nil, fmt.Errorf("given plaintext for encryption is nil")
```
(This is as close as GH's review tool lets me get 😀)

Line 231 can just be: `if len(plaintext) == 0 { `

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -268,29 +284,71 @@ func (k *OCIKMSSeal) Decrypt(ctx context.Context, in *physical.EncryptedBlobInfo
+			metrics.IncrCounter(metricDecryptFailed, 1)
+			return nil, errwrap.Wrapf("error decrypting data: {{err}}", err)
+		}
+		plaintext, err = base64.StdEncoding.DecodeString(*output.Plaintext)
```
`decryptedKey` (or something) would be better, since `plaintext` is used later to me something else.

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -16,30 +16,33 @@ import (
 	// OCIKMSEnvelopeAESGCMEncrypt is when a data encryption key is generated and
 	// the data is encrypted with AESGCM and the key is encrypted with KMS
-	OCIKMSEnvelopeAESGCMEncrypt
+	OCIKMSEnvelopeAESGCMEncrypt = iota
```
You can remove this since you do not have multiple versions for your encryption algorithm.

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+	return &policy
+}
+
+func (k *OCIKMSSeal) getCurrentKeyVersion(managementClient *keymanagement.KmsManagementClient) (string, error) {
```
I don't think you need to pass in the management client to this function.

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+// Copyright © 2019, Oracle and/or its affiliates.
+package ocikms
+
+import (
```
Can you group standard library at the top and third party libraries split by a newline and in a second group?

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+
+	// Check and set OCI KMS crypto client
+	if k.cryptoClient == nil {
+		kmsCryptoClient, err := k.getOCIKMSCryptoClient()
```
There is a lot of duplicated logic with the encrypt method.  I think it would be much easier to just run a simple encrypt operation instead of duplicating the logic here.

vuhpham-oracle replied:
Could you please elaborate? What logic really duplicate here?

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+		return nil, errwrap.Wrapf("error encrypting data: {{err}}", err)
+	}
+
+	keyVersion, err := k.getCurrentKeyVersion(k.managementClient)
```
It may be worth noting in a comment that there is a potential timing issue if the key gets rotated between this operation and the operation above.

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+
+	// Check and set authTypeAPIKey
+	var err error
+	k.authTypeAPIKey = false
```
This config item doesn't seem to be necessary.  You should just use this behavior if the key is present, otherwise fall back to instance auth.

vuhpham-oracle replied:
API Key can be shared for other features on the same host too. I think having this can to give user the flexibility to use instance auth even when some other features use API Key? What do you think?

vuhpham-oracle replied:
(Sorry I think I forgot to hit Submit for the comments)
OCI SDK doesn't enforce the hierarchy for access. It only provides the API for client to choose whether they want to use user/instance principal. Could you tell me why the bools are hard to configure?

chrishoffman replied:
Makes sense.

chrishoffman commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+
+	requestMetadata := k.getRequestMetadata()
+	var plaintext []byte
+	switch in.KeyInfo.Mechanism {
```
There is only one mechanism, so this logic is not required.

kalafut commented on vault/seal/ocikms/ocikms.go:
```diff
@@ -0,0 +1,414 @@
+
+func (k *OCIKMSSeal) Encrypt(ctx context.Context, plaintext []byte) (*physical.EncryptedBlobInfo, error) {
+	defer metrics.MeasureSince(metricEncrypt, time.Now())
+	if plaintext == nil || len(plaintext) == 0 {
```
@chrishoffman I don't think the `len(plaintext)==0` is necessary, and was wondering if it can actually cause problems to error out here? Thinking if there is some seal-wrap edge case where 0 bytes are encrypted.

vuhpham-oracle commented:
@kalafut @chrishoffman I have resolved latest comments on snake_case convention. Could you take a look and see if you have any further concern?
